### PR TITLE
e2e: CSI Tests

### PR DESCRIPTION
**Pull request by lpabon**
**What this PR does / why we need it**:
This e2e test tests the CSI external attacher with a mock CSI plugin driver.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*: https://github.com/kubernetes/features/issues/178

**Special notes for your reviewer**:
* Tests features in https://github.com/kubernetes/features/issues/178
* Tests implementation of https://github.com/kubernetes/community/pull/1258
* Tests VolumeAttachment Object: https://github.com/kubernetes/kubernetes/pull/54463

**Release note**:
```release-note
NONE
```

### CODE REVIEW

**Comment by lpabon** [comment]
cc @erinboyd @jsafrane

**Comment by lpabon** [comment]
To test this using `cluster/kube-up.sh` you must enable the following:

```
export KUBE_RUNTIME_CONFIG="storage.k8s.io/v1alpha1=true"
export KUBE_FEATURE_GATES="CSIPersistentVolume=true
```

**Comment by msau42** [comment]
@lpabon to enable these tests in CI, you need to modify these two places:
https://github.com/kubernetes/test-infra/blob/master/jobs/config.json#L4110
https://github.com/kubernetes/test-infra/blob/master/jobs/env/ci-kubernetes-e2e-gci-gce-alpha-features.env

Note that you cannot enable these tests in GKE until someone from google enables the v1alpha group.

**Comment by lpabon** [comment]
@msau42 nice! so I should add a `skip if gke` statement?

**Comment by msau42** [comment]
you don't need a skip statement as long as you don't add it to the gke env/config files.

File: test/e2e/storage/csi_volumes.go [hunk 1]
```diff
@@ -0,0 +1,544 @@
...
+			By("verifying nothing is published")
+			output := f.ExecCommandInContainer(pod.GetName(), "mock-plugin",
+				"csc", "controller", "list-volumes", "-v", "0.1.0", "--endpoint", "unix://"+csiSocketPath)
+			Expect(strings.Contains(output, "csi-mock/dev")).ToNot(BeTrue())
+			Expect(strings.Contains(output, "csi-mock/mnt/mock")).ToNot(BeTrue())
+
+			By("creating a persistentvolume")
+			pv := externalAttacherPersistentVolume(cs, config, false, false, mockPluginName, mockPluginVolumeHandle)
+
+			By("creating a fake volume attacher")
```
Comment (msau42) [inline test/e2e/storage/csi_volumes.go]: shouldn't the attach detach controller be the one creating/deleting the AttachVolume object?

Comment (lpabon) [reply test/e2e/storage/csi_volumes.go]: Currently that is not the task of the external-attacher. Instead is part of the plugin code created by #54529

Comment (msau42) [reply test/e2e/storage/csi_volumes.go]: E2E tests are supposed to test from the user's POV.  If you're stubbing out different components, that is better suited for an integration or unit test.

Comment (lpabon) [reply test/e2e/storage/csi_volumes.go]: Yeah I agree, but this is the only way to test a ...mostly.. e2e due to csi components. I will add to this test the e2e for #54529

Comment (lpabon) [reply test/e2e/storage/csi_volumes.go]: Ok, I will add to this PR another commit for a full e2e test.

File: test/e2e/storage/csi_volumes.go [hunk 2]
```diff
@@ -0,0 +1,544 @@
...
+			},
+			{
+				APIGroups: []string{""},
+				Resources: []string{"nodes"},
+				Verbs:     []string{"get", "list", "watch"},
+			},
+			{
+				APIGroups: []string{"storage.k8s.io"},
+				Resources: []string{"volumeattachments"},
+				Verbs:     []string{"get", "list", "watch", "update"},
```
Comment (edisonxiang) [inline test/e2e/storage/csi_volumes.go]: Resource "nodes" need Verbs "update" from Line 423;
Resource "volumeattachments" need Verbs "delete" from Line 351.

Comment (jsafrane) [reply test/e2e/storage/csi_volumes.go]: This is policy for external-attacher started as a pod somewhere in this PR. It's not policy for the test itself.

Comment (edisonxiang) [reply test/e2e/storage/csi_volumes.go]: @jsafrane Got it.

**Comment by lpabon** [comment]
Thanks all. I will have an update with a full e2e update coming.

File: test/e2e/storage/csi_volumes.go [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+	. "github.com/onsi/ginkgo"
+)
+
+const (
+	csiExternalAttacherImage    string = "docker.io/k8scsi/csi-attacher:0.1"
+	csiExternalProvisionerImage string = "docker.io/k8scsi/csi-provisioner:0.1"
+	csiDriverRegistrarImage     string = "docker.io/k8scsi/driver-registrar"
+)
+
+func externalAttacherServiceAccout(
```
Comment (msau42) [inline test/e2e/storage/csi_volumes.go]: Account

File: test/e2e/storage/csi_volumes.go [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+				provisioner:  "csi-hostpath",
+				parameters:   map[string]string{},
+				claimSize:    "1Gi",
+				expectedSize: "1Gi",
+				nodeName:     node.Name,
+			}
+
+			claim := newClaim(t, ns.GetName(), "")
+			class := newStorageClass(t, ns.GetName(), "")
+			testDynamicProvisioning(t, cs, claim, class)
```
Comment (msau42) [inline test/e2e/storage/csi_volumes.go]: Any plans to also test mount/unmount, attach/detach?

Comment (lpabon) [reply test/e2e/storage/csi_volumes.go]: Right now, we would need a driver that supports attach/detach. But mount/unmount is tested by `testDynamicProvisioning()`.

Comment (lpabon) [reply test/e2e/storage/csi_volumes.go]: Actually, now that I think about it, the test does test it. It all depends on the driver, but the e2e test does send out all the CSI calls to attach/detach and mount/unmount. All the CSI driver we have right now do attach/deatch and mount/unmount on the `NodePublish/Unpublish` calls done from the kubelet to the CSI driver. Probably when we get CSI drivers for the cloud vendors, the test would exercise attach/detach done from the `external-attacher` side car container to the CSI driver, then mount/unmount would be done by the kubelet to the CSI driver.

So, tldr: Yes, both are tested.

Comment (lpabon) [reply test/e2e/storage/csi_volumes.go]: Now that I think about it, maybe I should change the text on line 227 to say something like: `should provision, attach/detach, then mount/unmount storage with a hostPath CSI driver`. Would that make sense?

File: test/e2e/storage/csi_volumes.go [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+			{
+				APIGroups: []string{"storage.k8s.io"},
+				Resources: []string{"storageclasses"},
+				Verbs:     []string{"get", "list", "watch"},
+			},
+		},
+	}
+
+	clusterRoleClient.Delete(role.GetName(), &metav1.DeleteOptions{})
+	err := wait.Poll(2*time.Second, 10*time.Minute, func() (bool, error) {
```
Comment (jsafrane) [inline test/e2e/storage/csi_volumes.go]: I'd appreciate a constant for `10*time.Minute`, as it is used on multiple times

**Review by jsafrane — COMMENTED** [review]
One minor nit, otherwise LGTM

File: test/e2e/storage/csi_hostpath.go [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+					},
+					VolumeMounts: []v1.VolumeMount{
+						{
+							Name:      "socket-dir",
+							MountPath: "/csi",
+						},
+					},
+				},
+				{
+					Name:            "driver-registrar",
```
Comment (saad-ali) [inline test/e2e/storage/csi_hostpath.go]: I see that you bundled the all the external components and the driver in to the same pod. Not a huge deal, but our recommendation is two pods: 1) no more then one replica containing external-attacher, external-provisioner, and csi-driver and 2) on every node a driver-registrar and csi-driver.

We can consider this configuration for additional E2E tests in the future.

**Review by saad-ali — APPROVED** [review]
/lgtm
/approve
